src/platform-aws.cpp

```diff
@@ -768,51 +769,101 @@ int platform_config_endpoint(struct fi_info *info, struct fid_ep* endpoint) {
+    if (ret < 0) {
+        value = -errno;
+        goto cleanup;
+    }
+
+    fp = fopen(node_guid_filename, "r");
+    if (fp == NULL) {
+        NCCL_OFI_WARN("Error opening file: %s", node_guid_filename);
+        value = -errno;
+        goto cleanup;
+    }
+
+    /* Read the GUID in its colon-separated format */
+    ret = fscanf(fp, "%x:%x:%x:%x", &group[0], &group[1], &group[2], &group[3]);
+    if (ret != 4) {
+        NCCL_OFI_WARN("GUID parsing failed, got %d fields, expected 4", ret);
+        value = -EIO;
+        goto cleanup;
+    }
+
+    /* Calculate which group and which byte within the group */
+    group_idx = 3 - (position / 2);    // Convert position to group (3,2,1,0)
+    byte_idx = position % 2;           // 0 for low byte, 1 for high byte
+
+    /* Extract the correct byte */
+    if (byte_idx == 1) {
+        value = (group[group_idx] >> 8) & 0xFF;  // High byte
+    } else {
+        value = group[group_idx] & 0xFF;         // Low byte
+    }
+
+    NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET, "GUID: %04x:%04x:%04x:%04x, position: %d, group: %d, byte: %d, value: 0x%02x",
+                  group[0], group[1], group[2], group[3], position, group_idx, byte_idx, value);
 
-	ret = asprintf(&node_guid_filename, "/sys/class/infiniband/%s/node_guid",
-		       info->nic->device_attr->name);
-	if (ret < 0) {
-		vf_idx = -errno;
-		goto cleanup;
-	}
-	fp = fopen(node_guid_filename, "r");
-	if (fp == NULL) {
-		NCCL_OFI_WARN("Error opening file: %s", node_guid_filename);
-		vf_idx = -errno;
-		goto cleanup;
-	}
+cleanup:
+    if (node_guid_filename) {
+        free(node_guid_filename);
+    }
+    if (fp != NULL) {
+        fclose(fp);
+    }
+
+    return value;
+}
 
-	/**
-	 * GUID is a 64-bit hex number with format:
-	 *
-	 * XXXX:XXXX:XXXX:XXXX
-	 *
-	 * The lowest 8 bits are the VF id.
-	 */
-	ret = fscanf(fp, "%*x:%*x:%*x:%*2x%2x", &vf_idx);
-	if (ret != 1) {
-		NCCL_OFI_WARN("GUID parsing failed, got %d, expected 1", ret);
-		vf_idx = -EIO;
-		goto cleanup;
-	}
+uint64_t get_node_unique_id(struct fi_info *info) {
+    std::hash<std::string> hasher;
 
-cleanup:
-	if (node_guid_filename) {
-		free(node_guid_filename);
-	}
-	if (fp != NULL) {
-		fclose(fp);
-	}
+    /* Get the instance ID using the helper function */
+    const char* instance_id = nccl_net_ofi_get_instance_id();
+    if (instance_id == NULL) {
+        NCCL_OFI_WARN("Failed to get EC2 instance ID");
+        return 0;
+    }
 
-	return vf_idx;
-}
+    /* Get the per-card PCI bus (bits 16-23 from GUID) */
+    int per_card_pci_bus = get_guid_byte(info, 2);
+
+    /* Create node-unique portion by hashing instance_id */
+    std::string combined = std::string(instance_id) + "_" + std::to_string(per_card_pci_bus);
+    size_t hash_value = hasher(combined);
+    uint64_t node_unique = hash_value & ((1ULL << 56) - 1);
+
+    /* Get the func_idx (lowest 8 bits from GUID) */
+    unsigned int func_idx = (unsigned int)get_guid_byte(info, 0) & 0xFF;
```

Comment:
We should _not_ take the func_idx into account. That defeats the purpose of the GUID.